File: docs/content/geometry/convex-hull/code/convex_hull.v1.full.py
"""Convex hull — Andrew's monotone chain.

Computes the **convex hull**: the smallest convex polygon that contains a set of
2D points (imagine a rubber band snapping around them). Points are ``(x, y)``
tuples. Points lying *on* a hull edge (collinear) are dropped, so only the corner
vertices remain; the result is returned in counter-clockwise order.

Complexity: O(n log n) time (dominated by the sort), O(n) space.
"""


def cross(o: tuple[int, int], a: tuple[int, int], b: tuple[int, int]) -> int:
    """Cross product of vectors OA and OB.

    Args:
        o, a, b (tuple[int, int]): 2D points.
    Returns:
        int: > 0 if O->A->B turns left (CCW), < 0 if it turns right, 0 if collinear.
    """
    return (a[0] - o[0]) * (b[1] - o[1]) - (a[1] - o[1]) * (b[0] - o[0])


def convex_hull(points: list[tuple[int, int]]) -> list[tuple[int, int]]:
    """Convex hull of a set of points (Andrew's monotone chain).

    Args:
        points (list[tuple[int, int]]): the input points.
    Returns:
        list[tuple[int, int]]: the hull vertices, in counter-clockwise order.
    """
    # Sort by (x, y) and drop duplicates; the sweep needs a left-to-right order.
    pts: list[tuple[int, int]] = sorted(set(points))
    if len(pts) < 3:
        return pts                       # 0-2 distinct points: nothing to build

    # Lower hull: sweep left -> right, keeping only left turns. We pop the last
    # vertex while the edge does not turn left (right turn or collinear point).
    lower: list[tuple[int, int]] = []
    for p in pts:
        while len(lower) >= 2 and cross(lower[-2], lower[-1], p) <= 0:
            lower.pop()
        lower.append(p)

    # Upper hull: the same sweep in reverse builds the other half.
    upper: list[tuple[int, int]] = []
    for p in reversed(pts):
        while len(upper) >= 2 and cross(upper[-2], upper[-1], p) <= 0:
            upper.pop()
        upper.append(p)

    # Concatenate; drop each chain's last point (it repeats the other's start).
    return lower[:-1] + upper[:-1]

File: docs/content/geometry/convex-hull/code/convex_hull.v1.full.cpp
/**
 * Convex hull — Andrew's monotone chain.
 *
 * Computes the convex hull: the smallest convex polygon that contains a set of
 * 2D points (imagine a rubber band snapping around them).
 *
 * Input:  pts — a vector of 2D points (P{x, y}).
 * Output: the hull vertices in counter-clockwise order. Points lying *on* a hull
 *         edge (collinear) are dropped by the `<= 0` turn test, so only the
 *         corner vertices remain.
 * Complexity: O(n log n) time (dominated by the sort), O(n) space.
 */
#include <vector>
#include <algorithm>
using namespace std;

struct P {
    long long x, y;                          // 64-bit: cross() multiplies deltas
    bool operator<(const P& o) const {       // order the sweep by x, then by y
        return x < o.x || (x == o.x && y < o.y);
    }
};

// Cross product of vectors OA and OB. Its sign is the orientation of O->A->B:
//   > 0 left turn (counter-clockwise), < 0 right turn (clockwise), = 0 collinear.
long long cross(const P& O, const P& A, const P& B) {
    return (A.x - O.x) * (B.y - O.y) - (A.y - O.y) * (B.x - O.x);
}

vector<P> convex_hull(vector<P> pts) {
    int n = pts.size();
    if (n < 3) return pts;                    // 0-2 points: nothing to build
    sort(pts.begin(), pts.end());             // left-to-right sweep order

    vector<P> hull;                           // grows/shrinks like a stack
    // Lower hull: sweep left -> right, keeping only left turns.
    for (int i = 0; i < n; i++) {
        // Drop the last vertex while the edge does not turn left (right or
        // collinear): such a vertex lies inside the hull or on an edge.
        while (hull.size() >= 2 &&
               cross(hull[hull.size() - 2], hull.back(), pts[i]) <= 0)
            hull.pop_back();
        hull.push_back(pts[i]);
    }
    // Upper hull: sweep right -> left. `base` freezes the finished lower hull so
    // its vertices are never popped; pts[n-1] is skipped (already the last one).
    size_t base = hull.size() + 1;
    for (int i = n - 2; i >= 0; i--) {
        while (hull.size() >= base &&
               cross(hull[hull.size() - 2], hull.back(), pts[i]) <= 0)
            hull.pop_back();
        hull.push_back(pts[i]);
    }
    hull.pop_back();     // the last point repeats the starting vertex — drop it
    return hull;
}
